### PR TITLE
feat: Introduce FileUpload Question

### LINKED ISSUE #1257: [FEATURE][FormTribe 🔥][1000 Points] Question Type: File Upload
### Welcome to the FormTribe

This issue will get you the points mentioned in issue title for the [FormTribe Hackathon](https://formtribe.com)

**Preparation:**
If this is your first time contributing, please read:

- Our [introductory blog post](https://formbricks.com/blog/join-the-formtribe)
- Overview of [Code Guidelines at Formbricks.](https://formbricks.com/docs/contributing/how-we-code)

**Only when your PR follows _all_ of the best practices it will get merged.**

Want more points quick? Complete [Side Quests 🏰 ](https://formbricks.com/formtribe#side-quests)

--------

### Is your feature request related to a problem? Please describe.

Users want to enable respondents to upload files

### Describe the solution you'd like

Add question type "File Upload".

### Tasks

1. Add a new question type called File Upload with Hero Icon "ArrowUpTray": 
image

 The file type has the following settings:
image

The UI component for the toggle + card can be taken from the Survey Settings page

The Combobox component can be taken from the Filter:

image

2. Add a new component to the survey package to display file upload questions in surveys. The following mockups don't have to be rebuilt pixel-perfect, the font-size, colors etc are predetermined in the survey package - please stick to those:

image

3. Add logic: This new question type requires specific logic rules:

- Is empty
- Is not empty

Since this is binary, you can look at the logic rules for question type Consent. Make sure to adjust the logic options dependent on the Required toggle. If question is required, it should not be possible to select "Is empty" for a logic rule.

4. Test extensively :)

If you run into questions, please reach out to @pandeymangg on Discord - he is our Upload Guy 😎 

### Thanks!

### PR DESCRIPTION
**Pull request by KapadiaNaitik**
## What does this PR do?
Add File Upload Question

Fixes #1257  (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change adds a new database migration
- [ ] This change requires a documentation update

## How should this be tested?

- Create new question from the survey editor

### Required
`Will Update During Final Review`
- [ ] Filled out the "How to test" section in this PR
- [ ] Read [How we Code at Formbricks](<[https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md](https://formbricks.com/docs/contributing/how-we-code)>)
- [ ] Self-reviewed my own code
- [ ] Commented on my code in hard-to-understand bits
- [ ] Ran `pnpm build`
- [ ] Checked for warnings, there are none
- [ ] Removed all `console.logs`
- [ ] Merged the latest changes from main onto my branch with `git pull origin main`
- [ ] My changes don't cause any responsiveness issues

### Appreciated

- [ ] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

**Comment by jobenjada** [comment]
Hey @KapadiaNaitik - thanks for shipping this :)

Here is my feedback:

1. File upload field size changing on "Allow multiple files"

Off
image


On
image

--------

2. This is custom coded, but we already have a similar component in our settings cards:
image

For example this one:
image

For the means of consistency, please reuse the code form the settings card. We have a reusable input field which packs all the styling which is used there.

The text in the gray box should be: **Limit upload file size to [    ] MB.**

-------

3. File types settings

This isn't very clear. Do people add just the suffix or include the dot? Does it work if they do either? Is this a security risk for people to upload files with e.g. .js or .php etc?

We can tweak the UX of this later on, for now please just validate the input field to make sure people add max 4 letters and don't include the dot, any other symbols etc. Display senseful toasts if people try to input something which doesnt match the validation criteria:

image
 

-------

4. Single Response Card Display

Pls reuse the UI from the Summary Overview :)

image

-------

5. In Summary: Download instead of display

When clicking on this file, it leads to previewing the file in the browser. Instead, it should download it:


--------

4. Logic Jumps missing

@pandeymangg can you add basic logic jumps? E.g. has uploaded file / has not uploaded file?

-----

5. Email Preview missing

If File Upload is the first question and the user hits on the Email Preview tab in the Share Link Survey modal, it will likely break.

@pandeymangg can you add the corresponding component to the email preview functionality?

File: packages/types/product.ts [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
 import { ZColor, ZPlacement } from "./common";
+import { ZTeamBilling } from "./teams";
 
 export const ZProduct = z.object({
   id: z.string().cuid2(),
   createdAt: z.date(),
   updatedAt: z.date(),
   name: z.string(),
   teamId: z.string(),
+  billingInfo: ZTeamBilling.optional(),
```
Comment (mattinannt) [inline packages/types/product.ts:12]: please remove this from the product service as we can easily get this from the team service and we don't want to query an additional table every time we request a product.

File: packages/types/surveys.ts [hunk 2]
```diff
@@ -206,6 +215,16 @@ const ZSurveyQuestionBase = z.object({
   isDraft: z.boolean().optional(),
 });
 
+export const ZSurveyFileUploadQuestion = ZSurveyQuestionBase.extend({
+  type: z.literal(TSurveyQuestionType.FileUpload),
+  allowMultipleFiles: z.boolean(),
+  maxSizeInMB: z.number().optional(),
+  allowedFileTypes: z.array(ZAllowedFileExtension).optional(),
```
Comment (mattinannt) [inline packages/types/surveys.ts:222]: we sometimes have `fileExtensions` and sometimes `fileTypes`. Please choose one to be consistent 😊
I think `fileExtensions` is more common

File: packages/surveys/src/components/ui/FileInput/index.tsx
Comment (mattinannt) [inline packages/surveys/src/components/ui/FileInput/index.tsx:1]: the `ui` folder in components seem strange as we don't currently have this file convention in surveys package currently. Also in this PR were some attempts to change the component structure:
https://github.com/formbricks/formbricks/pull/916
Please check come up with a universal structure :-)

File: apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/components/FileUploadQuestionForm.tsx [hunk 3]
```diff
@@ -0,0 +1,235 @@
+import { TSurvey, TSurveyFileUploadQuestion } from "@formbricks/types/surveys";
```
Comment (mattinannt) [inline apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/components/FileUploadQuestionForm.tsx:1]: please add `"use client";`

**Review by mattinannt — CHANGES_REQUESTED** [review]

**Comment by jobenjada** [comment]
@mattinannt these things I couldn't solve:

1. Better error message when someone tries to upload a too large image. It just says "Upload failed!" - in case we can detect that it happened bc the file size is above the free level, we should display the File Too Large error message.

If not, we should at least add "Try again with a smaller file."

---------

2. When a file is deleted out of the uploader, the selection window is always opened again. I don't like the UX, bc when you delete something you dont necessarily want to upload a new file.

I've tried stopPropagation and moving the input a level up to not be a child but a sibling of the label (as GPT suggested). Can you fix this?

--------

Rest works well :)

**Comment by danielehrhardt** [comment]
> @mattinannt these things I couldn't solve:
> 
> 1. Better error message when someone tries to upload a too large image. It just says "Upload failed!" - in case we can detect that it happened bc the file size is above the free level, we should display the File Too Large error message.
> 
> If not, we should at least add "Try again with a smaller file."
> 
> 2. When a file is deleted out of the uploader, the selection window is always opened again. I don't like the UX, bc when you delete something you dont necessarily want to upload a new file.
> 
> I've tried stopPropagation and moving the input a level up to not be a child but a sibling of the label (as GPT suggested). Can you fix this?
> 
> Rest works well :)

Works great. Would like to see it in the next release <3

**Comment by pandeymangg** [comment]
The upload functionality is not working for in-app surveys right now, this is due to an overlook in the FileUpload component, will fix this today so that this can get closed 💪

File: packages/js/src/lib/widget.ts [hunk 4]
```diff
@@ -130,6 +130,19 @@ export const renderWidget = (survey: TSurvey) => {
...
       },
       onClose: closeSurvey,
+      onFileUpload: async (file: File, params) => {
+        const api = new FormbricksAPI({
+          apiHost: config.get().apiHost,
+          environmentId: config.get().environmentId,
+        });
+
+        const uploadedUrl = await api.client.storage.uploadFile(file, params);
+        console.log({
```
Comment (mattinannt) [inline packages/js/src/lib/widget.ts:140]: pleas remove

File: packages/surveys/src/components/general/FileInput.tsx [hunk 5]
```diff
@@ -0,0 +1,302 @@
+import { TAllowedFileExtension } from "@formbricks/types/common";
+import { useMemo } from "preact/hooks";
+import { JSXInternal } from "preact/src/jsx";
+import { useState } from "react";
+// import { uploadFile } from "../../lib/uploadFile";
```
Comment (mattinannt) [inline packages/surveys/src/components/general/FileInput.tsx:5]: please remove comment :-)

File: packages/surveys/src/components/general/FileInput.tsx [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+        const fileBuffer = await file.arrayBuffer();
+
+        const bufferBytes = fileBuffer.byteLength;
+        const bufferKB = bufferBytes / 1024;
+        if (bufferKB > maxSizeInMB * 1024) {
+          alert(`File should be less than ${maxSizeInMB} MB`);
+        } else {
+          setIsUploading(true);
+          try {
+            // const response = await uploadFile(file, allowedFileExtensions, surveyId, environmentId);
```
Comment (mattinannt) [inline packages/surveys/src/components/general/FileInput.tsx:44]: please remove comment :-)

File: packages/surveys/src/components/general/FileInput.tsx [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+            } else {
+              alert("Upload failed! Please try again.");
+            }
+          }
+        }
+      } else {
+        setIsUploading(true);
+
+        try {
+          // const response = await uploadFile(file, allowedFileExtensions, surveyId, environmentId);
```
Comment (mattinannt) [inline packages/surveys/src/components/general/FileInput.tsx:67]: please remove comment :-)

File: packages/surveys/src/components/general/FileInput.tsx [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+
+            const bufferBytes = fileBuffer.byteLength;
+            const bufferKB = bufferBytes / 1024;
+
+            if (bufferKB > maxSizeInMB * 1024) {
+              alert(`File should be less than ${maxSizeInMB} MB`);
+            } else {
+              setIsUploading(true);
+              try {
+                // const response = await uploadFile(file, allowedFileExtensions, surveyId, environmentId);
```
Comment (mattinannt) [inline packages/surveys/src/components/general/FileInput.tsx:135]: please remove comment :-)

File: packages/surveys/src/components/general/FileInput.tsx [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+                  alert(err.message);
+                } else {
+                  alert("Upload failed! Please try again.");
+                }
+              }
+            }
+          } else {
+            setIsUploading(true);
+            try {
+              // const response = await uploadFile(file, allowedFileExtensions, surveyId, environmentId);
```
Comment (mattinannt) [inline packages/surveys/src/components/general/FileInput.tsx:152]: please remove comment :-)

**Review by jobenjada — APPROVED** [review]

**Pull request merged**